sql/2020/07_SEO/lighthouse.sql

```diff
@@ -8,31 +8,68 @@ CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS
   ROUND(SAFE_DIVIDE(freq, total), 4)
 );
 
+CREATE TEMPORARY FUNCTION isCrawlableDetails(report STRING) 
+RETURNS STRUCT<
+disallow BOOL, 
+noindex BOOL,
+both BOOL,
+neither BOOL
+> DETERMINISTIC LANGUAGE js AS '''
+var result = {disallow: false, noindex: false};
+try {
+    var $ = JSON.parse(report);
+    var items = $.audits['is-crawlable'].details.items;
+    result.noindex = items.filter(item => item.source.type ==='node').length > 0;
+    result.disallow = items.filter(item => item.source.type ==='source-location').length > 0;
+} catch (e) {
+}
+return result;
+''';
+
 SELECT
     COUNT(0) AS total,
 
-    COUNTIF(is_crawlable) AS crawlable,
     COUNTIF(is_canonical) AS canonical,
-
-    AS_PERCENT(COUNTIF(is_crawlable), COUNT(0)) AS pct_crawlable,
     AS_PERCENT(COUNTIF(is_canonical), COUNT(0)) AS pct_canonical,
 
     COUNTIF(has_title) AS title,
-    COUNTIF(has_meta_description) AS meta_description,
-    COUNTIF(has_title AND has_meta_description) AS title_and_meta_description,
-
     AS_PERCENT(COUNTIF(has_title), COUNT(0)) AS pct_title,
+
+    COUNTIF(has_meta_description) AS meta_description,
     AS_PERCENT(COUNTIF(has_meta_description), COUNT(0)) AS pct_meta_description,
+
+    COUNTIF(has_title AND has_meta_description) AS title_and_meta_description, 
     AS_PERCENT(COUNTIF(has_title AND has_meta_description), COUNT(0)) AS pct_title_and_meta_description,
 
     COUNTIF(img_alt_on_all) AS img_alt_on_all,
     AS_PERCENT(COUNTIF(img_alt_on_all), COUNT(0)) AS pct_img_alt_on_all,
 
+    COUNTIF(link_text_descriptive) AS link_text_descriptive,
+    AS_PERCENT(COUNTIF(link_text_descriptive), COUNT(0)) AS pct_link_text_descriptive,
+
     COUNTIF(robots_txt_valid) AS robots_txt_valid,
     AS_PERCENT(COUNTIF(robots_txt_valid), COUNT(0)) AS pct_robots_txt_valid,
 
-    COUNTIF(link_text_descriptive) AS link_text_descriptive,
-    AS_PERCENT(COUNTIF(link_text_descriptive), COUNT(0)) AS pct_link_text_descriptive  
+    COUNTIF(is_crawlable) AS crawlable_pass,
+    AS_PERCENT(COUNTIF(is_crawlable), COUNT(0)) AS pct_crawlable_pass,
+
+    COUNTIF(is_crawlable_details.noindex) AS crawlable_noindex,
+    AS_PERCENT(COUNTIF(is_crawlable_details.noindex), COUNT(0)) AS pct_crawlable_noindex,  
+
+    COUNTIF(is_crawlable_details.disallow) AS crawlable_disallow,
+    AS_PERCENT(COUNTIF(is_crawlable_details.disallow), COUNT(0)) AS pct_crawlable_disallow,  
+
+    COUNTIF(is_crawlable_details.disallow AND is_crawlable_details.noindex) AS crawlable_disallow_noindex,
+    AS_PERCENT(COUNTIF(is_crawlable_details.disallow AND is_crawlable_details.noindex), COUNT(0)) AS pct_crawlable_disallow_noindex,  
```

Comment:
This is a bit nit picky, but is `crawlable_no_index` better phrased as `not_crawlable_no_index`? Or `crawlability_no_index`?

Reply:
This is where the lighthouse report got confused. noindex does not mean no crawl. In fact the bot has to crawl to see the noindex. disallow means it can't crawl. 

disallow_noindex is the case where crawling is blocked (disallow) and indexing is blocked (noindex).

Aleyda got confused by the crawlable_ prefix as it implies it represents only crawlable data. Maybe I drop the prefix?

Reply:
Will do. 

This is an area that confuses many SEOs. Google can index pages it is disallowed from crawling. And because it is disallowed Google does not see if a noindex is present. You sometimes see that in search results where the snippet does not show any content from the page, instead it shows a message saying it could not crawl.

e.g. a login page may be linked to from every page on a site, it's disallowed and noindexed. Google discovers it because of all the links, can't crawl it, so does not see the noindex. So it can show in search results with no information. In cases like this Google recommends letting them crawl to see the noindex.

This is why the authors wanted to see more details on what this "is_crawlable" actually meant.

Reply:
I think the naming is good now.